astropy/units/physical.py

```diff
@@ -1,133 +1,461 @@
+    property of `Unit` instances.  This class is not intended to be
+    instantiated directly in user code.
+
+    Parameters
+    ----------
+    unit : u.Unit
+        The unit to be represented by the physical type.
+
+    physical_types : `str` or `set` of `str`
+        A `str` representing the name of the physical type of the unit,
+        or a `set` containing strings that represent one or more names
+        of physical types.
+
+    Notes
+    -----
+    A physical type will be considered equal to an equivalent
+    `PhysicalType` instance (recommended) or a string that contains a
+    name of the physical type.  The latter method is not recommended
+    in packages, as the names of some physical types may change in the
+    future.
+
+    To maintain backwards compatibility, two physical type names may be
+    included in one string if they are separated with a slash (e.g.,
+    ``"momentum/impulse"``).  String representations of physical types
+    may include underscores instead of spaces.
+
+    Examples
+    --------
+    The preferred method to access a physical type is by accessing the
+    ``physical_type`` attribute of a unit.
+
+    >>> import astropy.units as u
+    >>> u.meter.physical_type
+    'length'
+
+    Some units correspond to multiple physical types.
+
+    >>> pressure = u.Pa.physical_type
+    >>> print(pressure)
+    {'energy density', 'pressure', 'stress'}
+    >>> 'energy density' in pressure
+    True
+
+    Physical types can be tested for equality against other physical
+    type objects or against strings that may contain the name of a
+    physical type.
+
+    >>> area = (u.m ** 2).physical_type
+    >>> area == u.barn.physical_type
+    True
+    >>> area == "area"
+    True
+
+    Multiplication, division, and exponentiation are enabled so that
+    physical types may be used for dimensional analysis.
+
+    >>> length = u.pc.physical_type
+    >>> area = (u.cm ** 2).physical_type
+    >>> length * area
+    'volume'
+    >>> area / length
+    'length'
+    >>> length ** 3
+    'volume'
+
+    Unknown physical types are labelled as ``"unknown"``.
+
+    >>> (u.s ** 13).physical_type
+    'unknown'
+
+    Dimensional analysis may be performed for unknown physical types too.
+
+    >>> length_to_19th_power = (u.m ** 19).physical_type
+    >>> length_to_20th_power = (u.m ** 20).physical_type
+    >>> length_to_19th_power, length_to_20th_power
+    ('unknown', 'unknown')
+    >>> length_to_20th_power / length_to_19th_power
+    'length'
+    """
+
+    def __init__(self, unit, physical_types):
+
+        if not isinstance(unit, core.UnitBase):
+            raise TypeError(f"Expecting a unit, not {unit}.")
+
+        self._unit = _replace_temperatures_with_kelvin(unit)
+        self._physical_type_id = self._unit._get_physical_type_id()
+        self._physical_type = _standardize_physical_type_names(physical_types)
+        self._physical_type_list = sorted(self._physical_type)
```

Comment:
Hmm, this makes me wonder a bit. It seems more reasonable to either leave it unsorted or keep user input (e.g., "pressure" is more common than "energy density", and both are more common than "shear").

Actually, I'm a bit confused why this is here; is it just for `__iter__`? I don't really see why that should be ordered.

Reply:
The main reason I'm sorting them is because [doctests require the line of output to match exactly](https://docs.python.org/3/library/doctest.html#warnings)...which can cause failures when the thing being printed out is a set or a string representation of a set.  This also happens when the physical types are printed out in error messages.  If the results are unordered, then we could run into situations where `str(physical_type) == str(physical_type)` returns `False`.

I like the idea of keeping the order that the items were inputted in, with the most common ones used first.

Reply:
Hmm, I don't really like that doctests drive design, but I can see how it is easier.

On second thought, I'm not sure any more I like my own suggestion of keeping order, but we're starting to risk overthinking it, so maybe best for you to just decide something and stick to it!

Reply:
In that case I'll stick with it being sorted, since it probably isn't going to make much difference one way or another when there are ≲ 4 physical type names, and most of the time there's only one name anyway.

That doctests are driving design is indeed irksome, though there's also the more general principle that it's harder to test code that is not deterministic...